zerver/migrations/0553_copy_emoji_images.py

```diff
@@ -0,0 +1,316 @@
+            )
+            print(f"{base_path}/{old_file_name} -> {base_path}/{new_file_name}")
+            try:
+                if os.path.exists(f"{base_path}/{new_file_name}.original"):
+                    os.unlink(f"{base_path}/{new_file_name}.original")
+                os.link(
+                    f"{base_path}/{old_file_name}.original", f"{base_path}/{new_file_name}.original"
+                )
+                with open(f"{base_path}/{new_file_name}.original", "rb") as fh:
+                    original_bytes = fh.read()
+            except OSError as e:
+                raise SkipImageError(f"Failed to read original file: {e}")
+
+            animated, still = resize_emoji(original_bytes, new_file_name)
+            try:
+                with open(f"{base_path}/{new_file_name}", "wb") as fh:
+                    fh.write(animated)
+
+                if still is not None:
+                    filename_no_extension = os.path.splitext(new_file_name)[0]
+                    with open(f"{base_path}/still/{filename_no_extension}.png", "wb") as fh:
+                        fh.write(still)
+            except OSError as e:
+                raise SkipImageError(f"Failed to write new file: {e}")
+
+            emoji.file_name = new_file_name
+            emoji.save(update_fields=["file_name"])
+        except SkipImageError as e:
+            logging.warning(
+                "Failed to re-thumbnail %s/emoji/images/%s: %s", emoji.realm_id, emoji.file_name, e
+            )
+            new_file_name = get_emoji_file_name("image/png", emoji.id)
+            try:
+                with open(f"{settings.DEPLOY_ROOT}/static/images/bad-emoji.png", "rb") as f:
+                    with open(f"{base_path}/{new_file_name}", "wb") as new_f:
+                        new_f.write(f.read())
+                emoji.deactivated = True
+                emoji.is_animated = False
+                emoji.file_name = new_file_name
+                emoji.save(update_fields=["file_name", "is_animated", "deactivated"])
+            except Exception as e:
+                logging.error("Failed to deactivate and replace with known-good image: %s", e)
+
+
+def thumbnail_s3(apps: StateApps) -> None:
+    total_processed = 0
+    avatar_bucket = boto3.resource(
+        "s3",
+        aws_access_key_id=settings.S3_KEY,
+        aws_secret_access_key=settings.S3_SECRET_KEY,
+        region_name=settings.S3_REGION,
+        endpoint_url=settings.S3_ENDPOINT_URL,
+        config=Config(
+            signature_version=None,
+            s3={"addressing_style": settings.S3_ADDRESSING_STYLE},
+        ),
+    ).Bucket(settings.S3_AVATAR_BUCKET)
+    for total_processed, emoji in enumerate(thumbnail_iterator(apps)):
+        if total_processed % 100 == 0:
+            print(f"Processed {total_processed} custom emoji")
+
+        old_file_name = emoji.file_name
+        try:
+            base_path = os.path.join(str(emoji.realm_id), "emoji/images")
```

Comment:
The S3 code path doesn't seem to have a check for if we've already processed this one.

Reply:
We need the content-type to build the new file name, which requires a network call -- so it's later, on line 237.